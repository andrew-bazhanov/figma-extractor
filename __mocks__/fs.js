import { fs } from 'memfs';

fs.__originalFs = jest.requireActual('fs');
module.exports = fs;
