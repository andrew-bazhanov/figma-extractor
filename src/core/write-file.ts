import fs from 'fs';
import path from 'path';

const fileLengthRuleDisableComment = '/* eslint-disable max-lines */';

const namingConventionRuleDisableComment =
  '/* eslint-disable @typescript-eslint/naming-convention */';

const defaultOptions = {
  useEslintDisabledRules: true,
};

// NOTE: temporarily changed writeFile to writeFileSync to make testing easier:
// it is unclear how to test callback version without introducing promises.
// But even with promises, the logic seems to "just write" contents without returning the promise itself,
// which is confusing a bit, because it is impossible to handle promise fulfillment.
export const writeFile = (
  template: string,
  filePath: string,
  options?: {
    useEslintDisabledRules?: boolean;
  },
) => {
  const dir = path.dirname(filePath);
  const useEslintDisabledRules = { ...defaultOptions, ...options }.useEslintDisabledRules;

  // Create directory in advance if there is no one
  if (dir && !fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }

  const content = useEslintDisabledRules
    ? `
    ${fileLengthRuleDisableComment}
    ${namingConventionRuleDisableComment}
    ${template}
    `
    : template;

  try {
    fs.writeFileSync(filePath, content);
    console.log(`Generate ${filePath} is completed`);
  } catch {
    console.log(`Formatting ${filePath} file...`);
  }
};
