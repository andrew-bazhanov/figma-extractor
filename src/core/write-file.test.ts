import fs from 'fs';
import { vol } from 'memfs';
import path from 'path';

import { writeFile } from './write-file';

jest.mock('fs');

describe('writeFile', () => {
  afterEach(() => {
    vol.reset();
  });

  it('checks existing of a new file', () => {
    expect(fs.existsSync('/ui/theme')).toBe(false);

    writeFile(
      `
    module.exports = {
      a: 1,
      b: 2,
      c: 3,
    };
    `,
      path.join('/ui/theme/', 'file-name.js'),
    );

    expect(fs.existsSync('/ui/theme')).toBe(true);
  });

  it('writes test data into a new file', () => {
    writeFile(
      `
    module.exports = {
      a: 1,
      b: 2,
      c: 3,
    };
    `,
      path.join('/ui/theme/', 'file-name.js'),
    );

    expect(vol.toJSON()).toMatchSnapshot();
  });
});
