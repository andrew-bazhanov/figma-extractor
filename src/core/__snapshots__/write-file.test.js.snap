// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writeFile writes test data into a new file 1`] = `
{
  "/ui/theme/file-name.js": "
    /* eslint-disable max-lines */
    /* eslint-disable @typescript-eslint/naming-convention */
    
    module.exports = {
      a: 1,
      b: 2,
      c: 3,
    };
    
    ",
}
`;

exports[`writeFile writes test data without eslint rules on the begin of the file 1`] = `
{
  "/ui/theme/file-name.js": "
    module.exports = {
      a: 1,
      b: 2,
      c: 3,
    };
    ",
}
`;
